Fingerprint unlock

ioana-farcas opened the pull request:
Fixes #68 

Todo:

- [x] handle callbacks ( go to list & ~fallback to PIN~ at #69)
- [x] add tests
- [x] address code review feedback
- [x] request UX review and address feedback (or separate issues for follow up)
- [x] add UI icon
- [x] use the DataProtect library for creating the CryptoObject #165

linuxwolf commented:
Note that implementing the fallback to PIN is #69.  We should be fine landing this without fallback, once the core "unlock with a fingerprint" is done

ioana-farcas commented:
@linuxwolf Thanks for letting me know

sashei commented on app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt:
```diff
@@ -0,0 +1,201 @@
+        fingerprintManager = manager
+    }
+
+    fun isFingerprintAuthAvailable(): Boolean {
```
We may want this to eventually be an `Observable<Boolean>` so that we can include the user-setting for fingerprint usage as well.

sashei replied:
I'm expecting this to be addressed in #31

jhugman replied:
👍

sashei commented on app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt:
```diff
@@ -0,0 +1,201 @@
+        }
+    }
+
+    private fun setupCiphers(): Pair<Cipher, Cipher> {
```
All of this key generation work can use the `DataProtect.Keystore` class instead of us doing the heavy lifting ourselves

sashei replied:
I'm going to make a follow-up story to address this because the right way to hook these up (observables? simple method calls?) is not obvious to me, and I'd rather get them both landed sooner than later.

jhugman replied:
:+1:

jhugman commented on app/src/main/java/mozilla/lockbox/view/LockedFragment.kt:
```diff
@@ -33,4 +33,9 @@ class LockedFragment : Fragment(), LockedViewProtocol {
+    override fun showFingerprintDialog() {
+        val dialogFragment = FingerprintAuthDialogFragment()
+        dialogFragment.show(fragmentManager, this.javaClass.simpleName)
+    }
```
This should be done from the `RoutePresenter`.

jhugman commented on app/src/main/java/mozilla/lockbox/view/FingerprintAuthDialogFragment.kt:
```diff
@@ -0,0 +1,99 @@
+
+    override fun onResume() {
+        super.onResume()
+        view!!.imageView.setImageResource(R.drawable.ic_fingerprint)
```
The need to reset this here will likely go away if you use a `PublishSubject` instead of a `ReplaySubject` for the authentication state.

jhugman commented on app/src/main/res/layout/fragment_fingerprint_dialog.xml:
```diff
@@ -0,0 +1,62 @@
+            app:layout_constraintTop_toTopOf="parent" android:layout_marginTop="44dp" android:textSize="17sp"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent" android:layout_marginStart="52dp"
+            android:layout_marginEnd="51dp"/>
```
Nit: The layout of these looks somewhat haphazard. Recommend we adopt the Android style in Android Studio.

ioana-farcas commented on app/src/main/java/mozilla/lockbox/action/AuthenticationAction.kt:
```diff
@@ -0,0 +1,14 @@
+sealed class AuthenticationAction : Action {
+    object StartListening : AuthenticationAction()
+    object StopListening : AuthenticationAction()
+}
```
Maybe the "AuthenticationAction" name does not fit. It would be better "FingerprintSensorAction", because the sensor starts and stops listening for fingerprint. What do you think?

devinreams commented:
@ioana-farcas I updated the "todo list" up above. Does that capture everything left?

I believe you said by this time tomorrow it will be ready for UX review so I re-requested reviews from the engineering and UX teams so its on their mind.

And is there anything we as a team can do to help finish this soon so other stories can be unblocked?

sashei commented on app/src/main/java/mozilla/lockbox/presenter/FingerprintDialogPresenter.kt:
```diff
@@ -0,0 +1,73 @@
+            .subscribe {
+                when (it) {
+                    is AuthCallback.OnAuth -> dispatcher.dispatch(RouteAction.ItemList)
+                    is AuthCallback.OnError -> dispatcher.dispatch(RouteAction.LockScreen)
```
awesome

ioana-farcas commented:
@devinreams There is one more item: using the DataProtect library for creating the CryptoObject.
I will finish the callbacks and addressing the code review comments so that this can be reviewed by the UX team. And I will ask James to help me with re-reviewing the changes, to have this done until the standup.
I am thinking to finish this part so that this can be merged in order to unblock the other tickets and maybe create another ticket for the DataProtect library item.

@jhugman, @sashei Thanks for the reviews! :)

sashei commented:
@ioana-farcas I made a follow-up issue for the `CryptoObject`: #165

jhugman commented on app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt:
```diff
@@ -0,0 +1,206 @@
+
+    init {
+        dispatcher.register.filterByType(FingerprintSensorAction::class.java)
+            .doOnDispose { stopListening() }
```
Nice.

jhugman replied:
I wander if we can do away with the `Start` and `Stop` actions if we can `doOnDispose()` and `doOnSubscribe()`.

Please file a `tech-debt`/`chore` issue.

jhugman replied:
Filed as #166.

No further action needed here.

jhugman commented on app/src/main/java/mozilla/lockbox/store/LockedStore.kt:
```diff
@@ -0,0 +1,24 @@
+    val onAuthentication: Observable<FingerprintAuthAction> =
+        dispatcher.register
+            .filterByType(FingerprintAuthAction::class.java)
+}
```
Yes!

jhugman commented on app/src/main/java/mozilla/lockbox/view/DialogFragment.kt:
```diff
@@ -0,0 +1,36 @@
+import android.view.View
+import mozilla.lockbox.flux.Presenter
+
+open class DialogFragment : DialogFragment() {
```
Hmm. I see this makes sense.

I should file a follow up to rename `CommonFragment` to `mozilla.lockbox.view.Fragment`.

ioana-farcas commented:
@nickbrandt 
- I'm trying to see if I can handle this error separately. Usually when an error is received that disables the sensor, the dialog is dismissed. I have to see if this is the case for this error
- The behaviour of the app when it's closed and then reopened is not part of this ticket
- There were some navigation issues when you verified the app, please try again and let me know if the app still crashes. It should be fine now :)
- Automatically prompting the user when returning to app is not part of this ticket
- I added the lockbox icon, please take a look when you have time
- The icons can be animated, but I have to look into this, animations in android are not straight forward. We could use a library for animations or we can implement them, I will investigate a little and come with an estimation

linuxwolf commented:
>     * I'm trying to see if I can handle this error separately. Usually when an error is received that disables the sensor, the dialog is dismissed. I have to see if this is the case for this error

From what I recall, the `onAuthenticationHelp()` callback is triggered if the error is recoverable.  As far as I can tell from a glance at the code, there is no differentiation between recoverable and unrecoverable errors.

>     * The icons can be animated, but I have to look into this, animations in android are not straight forward. We could use a library for animations or we can implement them, I will investigate a little and come with an estimation

I created a follow-up issue (#170) to account for this.

nickbrandt commented:
@ioana-farcas that all makes sense to me. The icon looks great, though it is appearing a bit off center (see screenshot). Once that is updated and if we can get it so that the prompt doesn't dismiss with that error case, I think we are good! Thanks for all your work on this, looks great! 👍 

![screenshot_20181025-090815](https://user-images.githubusercontent.com/5185090/47511069-a4d42400-d836-11e8-8390-a1175e4974dd.png)

ioana-farcas commented:
@nickbrandt I aligned the icon and handled the recoverable errors so the dialog will not be dismissed any more. Please take a look and thanks for the reviews :)

nickbrandt commented:
@ioana-farcas Looks and functions perfect! Thank you :)

linuxwolf commented:
Just tested the branch on my physical device (Pixel 3XL), and works beautifully.  Thansk @ioana-farcas !

:shipit: 🚀

devinreams commented:
@ioana-farcas great work! nicely done!

![200w_d](https://user-images.githubusercontent.com/49511/47520497-97766400-d84d-11e8-9930-efe58e9ccaee.gif)

sashei commented:
💯 🏅 🎉
